### ISSUE TITLE
[BUG] New salt version is incompatible with SEP 14

### ISSUE THREAD

**Issue by Foxboron** [saltstack/salt#63048]
**Description**

The release numbering of saltstack is documented in [SEP 14](https://github.com/saltstack/salt-enhancement-proposals/blob/master/accepted/0014-dev-overhaul.md). But this only describes a `MAJOR.MINOR` release numbering. The current release has a dash in it, `v3005.1-2`, and It's not clear what this should be interpret as.

It should be noted it will probably not work with a lot of downsteam semver'ish comparisons.

**Comment by Foxboron** [comment]
I think there is some confusion how upstream (the salt project) and downstream (debian, redhat, archlinux) handles the relative package version.

The relative version of package is only a concept in downstream packaging. This is to deal with rebuilds, patching or if the distribution forgets to include man pages. This is not something the upstream package should be dealing with. The version released by the upstream project here is still a version release.

This is the changelog of the release https://github.com/saltstack/salt/compare/v3005.1...v3005.1-2 and all of this is not packaging changes. These are upstream changes to deal with dependency requirements and should be released in a patch release like `v3005.2`.

If salt the company need to release a package change for the packages they provide, then this needs to be dealt with in the respective packages files, not the upstream project.

The issue now is that `v3005.1-2` is not going to be packaged in downstream distributions as `3005.1-2`. Most likely we will end up having to replace the dash with a dot to represent a patch release (`v30005.1.2`) as downsteams append their own relative package version.

The question then is if salt expects downstream to package `v3005.1-2` or not.

**Comment by OrangeDog** [comment]
The Salt Debian package is now `3005.1+ds-2`, for example.

I agree though, that ec1db3a should have made this `3005.2`. I didn't realise that change was in there.

**Comment by Foxboron** [comment]
>The Salt Debian package is now 3005.1+ds-2, for example.

From the packages salt provides, yes?

It does not make sense for the salt project to include the relative version in an upstream release. The version number you have linked says you have packaged version `3005.1` with a second repackaging of the release for version `2`. This is not the case. You have packaged the tarball from the `30005.1-2` and probably still on the first package iteration of this release.

This is the upstream repository and should not concern itself with downstream packaging. The fact that you are confusing this role is because you are your own upstream and your own downstream.

I'm not sure how to be more clear at this point.

**Comment by OrangeDog** [comment]
I believe the intent was for this to only have changed the dependencies, hence it was a repackaging rather than any code changes to the product.

**Comment by Foxboron** [comment]
The files where changed in the upstream project thus it should have been a minor release, or a patch release. A "repackaging" release is not something that should be happening in the upstream repository.

**Comment by cr1xu5** [comment]
It seems like the Debian package provided in https://repo.saltproject.io/salt/py3/ubuntu/22.04/amd64/latest/pool/ already was version `3005.1+ds-2` and now provides a changed package under exact the same name.

```
admin@salt ~ % apt list --upgradable | grep salt

WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

salt-common/unknown 3005.1+ds-2 all [upgradable from: 3005.1+ds-2]
salt-master/unknown 3005.1+ds-2 all [upgradable from: 3005.1+ds-2]
salt-minion/unknown 3005.1+ds-2 all [upgradable from: 3005.1+ds-2]
admin@salt ~ %
```

Could this have been the cause for problems with Python dependencies under Ubuntu 22.04? We have a bricked Salt master now:

```
admin@server1 ~ % sudo apt install python3-pygit2 python3-git
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
python3-git is already the newest version (3.1.24-1).
python3-pygit2 is already the newest version (1.6.1+dfsg-2).
0 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.
admin@server1 ~ % journalctl -u salt-master
...
Nov 09 08:29:03 server1 systemd[1]: Starting The Salt Master Server...
Nov 09 08:29:06 server1 systemd[1]: Started The Salt Master Server.
Nov 09 08:29:06 server1 salt-master[1187264]: [ERROR   ] gitfs is configured but could not be loaded, is GitPython installed?
Nov 09 08:29:06 server1 salt-master[1187264]: [CRITICAL] No suitable gitfs provider module is installed.
Nov 09 08:29:06 server1 salt-master[1187264]: [CRITICAL] No suitable git_pillar provider module is installed.
Nov 09 08:29:06 server1 salt-master[1187264]: [CRITICAL] Failed to load git_pillar
Nov 09 08:29:06 server1 salt-master[1187264]: [CRITICAL] Master failed pre flight checks, exiting
Nov 09 08:29:06 server1 systemd[1]: salt-master.service: Main process exited, code=exited, status=1/FAILURE
Nov 09 08:29:06 server1 systemd[1]: salt-master.service: Failed with result 'exit-code'.
Nov 09 08:29:06 server1 systemd[1]: salt-master.service: Consumed 2.590s CPU time.
...
admin@salt2 ~ %
```

**Comment by dmurphy18** [comment]
This is purely a packaging artifact from the GitLab CI/CD used to build the onedir packaging.  There are actually no changes apart from the bump in '-2' for the release number.  The actual changes occurred in the RedHat family of packages due to additions in the 'post minion' and 'post master' and their postun equivalents in the spec file.  No actual changes in Salt, hence the Release number bump.

Effectively for Debian / Ubuntu this was a rebuild and Release number bump.
There was an associated classic Windows packaging change to pick up a newer version of pyzmq and it got the same Release number bump, but that is classic packaging and Windows, so not related here.

The versioning is 3005.1-2 for Redhat family, 3005.1+ds-2 for Debian / Ubuntu, changes are purely packaging, and for Debian / Ubuntu due to GitLab CI/CD tools, cannot separate out the platforms, it builds them all, Linux, Windows, MacOS, etc.

@Foxboron given this explanation, please consider closing this issue, unless you need further explanation.

**Comment by Foxboron** [comment]
It doesn't explain what you are expecting other downstreams to do with the release, and if this is going to be a common occurrence going forward.

The packaging releases and the upstream releases should be decoupled, and the explanation given still does not really give justify mixing up packaging releases and upstream releases.

If this is a limitation in how you structure packaging building then I'd suggest you try fix that, else you should still be tagging this as minor releases or patch releases.

**Comment by dmurphy18** [comment]
@Foxboron However you raise an interesting point about https://github.com/saltstack/salt-enhancement-proposals/blob/master/accepted/0014-dev-overhaul.md , but it's date 2019-10-01, show it predates almost use of onedir.  Onedir was being using to built native minions for Arista routers at this point, and the builds were done by hand, but the SEP does not consider Release numbers, that is, packaging number bumps, as part of the version, since they are not related to Salt versioning but packaging, and only change when packaging changes occur.

The difference here is with current CI/CD tooling, need to provide Release number as part of the tag to have the correct Release number used in the build process, as this allowed revised packaging tests to also be used, again, totally unrelated to Salt versioning.  For example: when 3005.1-1 was produced, Windows was using Python 3.8.14, but currently Windows is using Python 3.8.15, hence tests had to be updated to compare against Python 3.8.15.  What a difference a month makes :) .

Tried to work-around this in the CI/CD (two days later), but the only way to resolve issues was with the tag.

Moving forward, and this is after much discussion in the core-team, the tagging will have to include the Release number, since now including dependencies in onedir, which are not related to the Salt version actually used.

Hypothetically, Salt Release 3020.2-1, and have to update pyzmq to a CVE, then we have 3020.2-2, but the version of Python is updated due to a CVE, we get 3020.2-3, etc.

The Release number bump occurs due to packaging changes and not changes in Salt.
This problem is not confined to onedir, for example: the classic packaging for Salt 3005 on Ubuntu 20.04 also got a Release number bump since Salt had to provide a version of pyzmq which does not have concurrency issues, which the OS supplied pyzmq v18.1.1 has.
```
py3/ubuntu/20.04/amd64/archive$ l
total 44K
drwxr-xr-x 11 ubuntu ubuntu 4.0K Sep 29 17:21 .
drwxr-xr-x  3 ubuntu ubuntu 4.0K Sep 29 17:44 ..
drwxr-xr-x  6 ubuntu ubuntu 4.0K Mar 22  2022 3002.8
drwxr-xr-x  6 ubuntu ubuntu 4.0K May 31 20:01 3002.9
drwxr-xr-x  6 ubuntu ubuntu 4.0K Mar 16  2022 3003.4
drwxr-xr-x  6 ubuntu ubuntu 4.0K May 25 20:39 3003.5
drwxr-xr-x  6 ubuntu ubuntu 4.0K Mar  8  2022 3004.1
drwxr-xr-x  6 ubuntu ubuntu 4.0K May 16 19:25 3004.2
lrwxrwxrwx  1 ubuntu ubuntu    6 Sep 14 20:49 3005 -> 3005-2
drwxr-xr-x  6 ubuntu ubuntu 4.0K Aug 23 17:32 3005-1
drwxr-xr-x  6 ubuntu ubuntu 4.0K Sep  1 20:02 3005-2
drwxr-xr-x  6 ubuntu ubuntu 4.0K Sep 29 17:26 3005.1
py3/ubuntu/20.04/amd64/archive$
```

**Comment by dmurphy18** [comment]
@Foxboron Was finishing up my comment when you replied. Hopefully this answers some of the issues you have, and actually shows reasoning for the coupling.  The problem is due to onedir not being similar to OS packaging formats where the package does the minimum and utilizes other dependency packages rather than all-inclusive.  However this has proven to be a problem for Salt, and I have spent much more time fixing dependency packaging for an OS platform than I have packaging Salt, for example: 
* problem with pyzmq on Ubuntu 20.04
* RedHat continue to use EOL Python 3.6 on RHEL 7 and 8
* Debian 9 with Python 3.5 preventing the use of f-strings
* OS dependencies with CVE's which are not being fixed

The solution was to bring those dependencies into a self-contained area with Salt, so that Salt could have more control over what it uses, and allows Salt to function on other platforms which do not have modern support, that is, only lang 'C' and Python 2.7, and UTF-8 and Python 3 will be a long time coming.

Just saw your new comment and 
```
They will need to append their own packaging number because - is usually an illegal character for version numbers in downstream package managers.
```
'-' is used frequently in packaging, so not understanding the point you are trying to make here, other than the stripping of '-', noting the Release bump and other characters such as '~', bpo, etc do complicate release bumps, so it is not as if this area isn't without its complications.  But understand, the package managers are adding theirs and not expecting it in the tag, typically.

The core team has discussed doing 3020.2.3 instead of 3020.2-3 too, and the SEP will need updating given it is over 3 years old, and doesn't contain any of these issues, that will occur once a final (for now :) ) decision is reached. I will bring up the issue within the core-team during our next standup.

**Comment by Foxboron** [comment]
>'-' is used frequently in packaging, so not understanding the point you are trying to make here, other than the stripping of '-', noting the Release bump and other characters such as '~', bpo, etc do complicate release bumps, so it is not as if this area isn't without its complications. But understand, the package managers are adding theirs and not expecting it in the tag, typically.

The point I'm trying to make is that even if you are internally making a distinction between `v3005.1-2` and `v3005.1` there is no actual differences. They are two tags in the git repository marking a release.

Because of this you will see people packaging this release and dotting the dash as it makes no sense for downstream to follow to your repackaging.

**Comment by dmurphy18** [comment]
I got that

**Comment by waynew** [comment]
There will certainly be some more discussion but just to note: a tag is not necessarily a release.

It frequently is, but there's nothing emergent about a tag that requires it to be a release.

In fact, tags *should* work the other way around. Instead of saying "this tag creates a release" it should be said "we created a relase from this revision, and are tagging it to make it easier to find" (and whatever other reasons)

**Comment by Foxboron** [comment]
>There will certainly be some more discussion but just to note: a tag is not necessarily a release.

At some point in November I was going to point out how you did cut releases on github with those tags;
https://github.com/saltstack/salt/releases/tag/v3005.1-2

However this seems to have stopped after that initial release, so that is good.

If there isn't any new progress on this I'll close it to tidy up my open issues.

**Comment by dmurphy18** [comment]
@Foxboron Can you take a look at the just released Salt 3006.0, and see if we have resolved your issues.
Moved to completely new CI/CD using GitHub Actions, and not using the pipeline on GitLab CICD anymore, and hopefully with the switch to GitHub Actions, we have resolved the issues you raised.

If so, please consider closing this issue.
